### ISSUE TITLE
Problem with Snapping

### ISSUE THREAD

**Issue by Fatih20** [xournalpp/xournalpp#1972]
Grid Snapping does nothing, when moving objects. I've been told that it enables snapping when objects are being moved to be constrained to an invisible grid in the document. Yet, as I demonstrated  in the attached video below, it didn't snap.

https://open.lbry.com/Screen-record-from-2020-05-12-19.12.20#9ce4e31b5bcd0ebc814b2443dd82469e73403834

The solution to this seems to be pretty obvious and I hope it can get fixed because it's a bad user experience to have a feature not working.

**Comment by Febbe** [comment]
You should really work on your temper. Currently you are rude and unfriendly. In general this will lead to people not interested to help you out with that, or generally to deal with you. Also it will be very embarrassing if your problem was an RTFM or layer 8 problem.
Also you should stick to our bug template, currently I don't know if this could be a new bug, or a already fixed one. 

As first hint you should upgrade and then visit the preferences first. There are some additional options now. Also rewrite your issue if it still persists.

**Comment by Fatih20** [comment]
> You should really work on your temper. Currently you are rude and unfriendly. 
Sorry, I was just using this app and got really frustrated about this problem.
> RTFM or layer 8 problem.
The user guide and the FAQ doesn't say anything about snapping, https://github.com/xournalpp/xournalpp/wiki/User-Manual#snapping
> As first hint you should upgrade and then visit the preferences first. There are some additional options now. Also rewrite your issue if it still persists.
I'm using the version available through the ppa. I did decrease snapping tolerance but it stayed the same when moving object. Effects of grid snapping can only be felt when making new object.

**Comment by Febbe** [comment]
> > RTFM or layer 8 problem.

 > The user guide and the FAQ doesn't say anything about snapping, https://github.com/xournalpp/xournalpp/wiki/User-Manual#snapping

Oh clicking on that link is really frustrating. Expected just old infos, not empty chapters with an header.
Also this was just an example, bad behavior everytime falls back on your feet and even harder, if you aren't right.

Thanks for the additional informations. The work on the snapping function is still in progress, but it got already massive improvements the last few weaks. And I am aware of the problem, that some edit tools just don't respect the snapping.

It would be a great help, if you edit your issue and summarize all the tools which are currently affected.

**Comment by rolandlo** [comment]
What exactly is the expected snapping behaviour when moving a stroke, an image or another element? What point should be snapped? This seems not quite obvious. In the code there is only rotation snapping for moving objects, no snapping to grid (see EditSelection class).

In any case this issue is a duplicate of #807.

**Comment by Fatih20** [comment]
> What exactly is the expected snapping behaviour when moving a stroke, an image or another element? What point should be snapped? This seems not quite obvious. In the code there is only rotation snapping for moving objects, no snapping to grid (see EditSelection class).
> 
> In any case this issue is a duplicate of #807.

Example from Inkscape.
https://open.lbry.com/Screen-record-from-2020-05-13-03.52.55#02623c2eae5cf64b94217405ad474defa5707467

**Comment by rolandlo** [comment]
Thanks @Faith20 for the Inkscape example. As I see the snapping behaviour of moving objects is defined by the corner of the bounding box which is closest to the cursor position. That kind of makes sense. We can implement this behavious in xournal++ as well. That's no big deal.

**Comment by frohro** [comment]
I'm not sure the corner of the bounding box is just what we want.  I think we probably want the corner of the object in the bounding box closest to where it was grabbed.  Maybe this is what you meant.

**Comment by frohro** [comment]
I think this is a duplicate of issue #807.

**Comment by rolandlo** [comment]
Yes @frohro, that's basically what I meant. For a stroke I would just compute the minimal and maximal x-/y-coordinates xMin, xMax, yMin, yMax among all points of the stroke and take the point from (xMin, yMin), (xMin, yMax), (xMax, yMin) and (xMax, yMax) nearest to the point where the selection was grabbed. This is a similar computation like in the calcSize() function, just without margin and pen thickness.

By the way I already pointed out the duplicate issue.

**Comment by LittleHuba** [comment]
And those points are exactly the corners of the bounding box 😉

**Comment by frohro** [comment]
Is that really right?  I must not understand.  Here is what it looks like to me.  The bounding box is some  distance away from the stroke that seems independent of the stroke width.  See the photos for my confusion.
![2020-05-14_16-35](https://user-images.githubusercontent.com/994565/81996057-ec7cf500-9600-11ea-9293-7c61eefda562.png)
![2020-05-14_16-33](https://user-images.githubusercontent.com/994565/81996060-ed158b80-9600-11ea-9da6-7965d6503ed1.png)

**Comment by rolandlo** [comment]
> And those points are exactly the corners of the bounding box minus stroke width upside_down_face

I disagree on the word "exactly". The bounding box (boundingRect() in the code) uses the calcSize() method. In it there is also some margin (2 points to each side). Moreover the "stroke width" is not constant if the stroke has been painted with pressure. It depends on each point. Take a look at the code snipped (taken from Stroke::calcSize()):
```

    bool hasPressure = points[0].z != Point::NO_PRESSURE;
    double halfThick = this->width / 2.0;  //  accommodate for pen width

    for (auto&& p: points) {
        if (hasPressure) {
            halfThick = p.z / 2.0;
        }

        minX = std::min(minX, p.x - halfThick);
        minY = std::min(minY, p.y - halfThick);

        maxX = std::max(maxX, p.x + halfThick);
        maxY = std::max(maxY, p.y + halfThick);
    }

    Element::x = minX - 2;
    Element::y = minY - 2;
    Element::width = maxX - minX + 4;
    Element::height = maxY - minY + 4;
```
So because of variable pressure it is not quite enough to use the corners of the bounding box and add/subtract some constant number. It must be recomputed, I think.

**Comment by rolandlo** [comment]
> I have been using your snap to grid splines to make inductors nicely. Thanks for that great fix!

Wonderful that it is already of some good use.

**Comment by LittleHuba** [comment]
Ahh taking all back pressure sensitivity does make it awkward 😅 the additional margin should not be part of the bounding box though. But I have no idea why it is there to start with...

**Comment by frohro** [comment]
Interesting investigation of the code @rolandlo!  I wonder if we could use the corners of the box you display above as the positions to snap to, but keep the extra 3 points so you move the whole object....  What do you think?  I would personally like the middle of the line to snap to the grid, but free hand lines snapping to grid is not nearly as important as the objects, like the square, line, etc. and those don't seem to have pressure sensitivity, so even this would be a great improvement.

**Comment by rolandlo** [comment]
Well, my idea is to take the corner nearest to the position the selection has been grabbed, compute the corresponding snapped point, compute the differences dx, dy to the original position and then translate the whole selection by (dx,dy).

**Comment by frohro** [comment]
I like that.

**Comment by rolandlo** [comment]
This has been fixed in PR #2011
